sycl/doc/SYCL2020-SpecializationConstants.md

```diff
@@ -0,0 +1,874 @@
+intermediate representation is used), we need to somehow emulate support for
+specialization constants.
+
+## Design
+
+As stated above, native specialization constants support is based on
+corresponding SPIR-V functionality, while emulation is supposed to be
+implemented through transforming specialization constants into kernel arguments.
+
+In DPC++ Headers/DPC++ RT we don't know a lot of necessary information about
+specialization constants, like: which numeric ID is used for particular
+specialization constant (since we support `SYCL_EXTERNAL`, those IDs can only
+be allocated by the compiler during link stage) or which kernel argument is used
+to pass particular specialization constant (because they are not explicitly
+captured by SYCL kernel functions and regular mechanism for kernel arguments
+handling can't be used here).
+
+Therefore, we can't have headers-only implementation and the crucial part of
+design is how to organize mapping mechanism between SYCL identifiers for
+specialization constants (`specialization_id`s) and low-level identifiers
+(numeric IDs in SPIR-V or information about corresponding kernel arguments).
+
+That mapping mechanism is particularly tricky, because of some additional
+complexity coming from SYCL 2020 specification:
+- `specialization_id` variables, which are used as specialization constant
+  identifiers (being non-type template parameters of some methods) can't be
+  forward-declared in general case (for example, if defined as `static`), which
+  means that we can't use integration header to attach some information to them
+  through some C++ templates tricks (like it is done for regular kernel
+  arguments or kernel names, for example).
+- they also can be declared as `static` or just non-`inline` `constexpr`, which
+  means that they have internal linkage and can't be referenced from other
+  translation units, which means that we can't for example create a new
+  translation unit which contains some mapping from `specialization_id` address
+  to some desired info.
+
+Based on those limitations, the following mapping design is proposed:
+- DPC++ RT uses special function:
+  ```
+  namespace detail {
+    template<auto &SpecName>
+    inline const char *get_spec_constant_symbolic_ID();
+  }
+  ```
+  Which is only declared, but not defined in there and used to retrieve required
+  information like numeric ID of a specialization constant.
+- Definition of that function template are provided by DPC++ FE in form of
+  _integration footer_: the compiler generates a piece of C++ code which is
+  injected at the end of a translation unit:
+  ```
+  namespace detail {
+    // assuming user defined and used the following specialization_id:
+    // constexpr specialiation_id<int> int_const;
+    // class Wrapper {
+    // public:
+    //   static constexpr specialization_id<float> float_const;
+    // };
+
+    template<>
+    inline const char *get_spec_constant_symbolic_ID<int_const>() {
```

Comment:
So... how does this line work in situations where we have global-linkage but aren't allowed to spell the specialization constant's name?  For example, if it is also a local variable inside a function?

```
void foo() {
  static specialization_id<float> f;
  thread_local specialization_id<int> f2;
}
```
OR, what if it is a 'private' static variable?  We cannot spell that either.

Reply:
Another fun one!  Anonymous namespaces!  Which 'i' do we mean in this example?  Which can we do this specialization with?

```
namespace ParentNS {
specialization_id<int> MySpecID;
namespace {
specialization_id<int> MySpecID;
}
};
// Both are spelled as ParentNS::MySpecID.  We can't actually do BOTH.  Language rules show we would only be able to name the one in ParentNS, not the one in the anonymous namespace.
```


```
namespace ParentNS {
specialization_id<int> MySpecID;
namespace {
specialization_id<int> MySpecID2;
}
};
// However, if they have different names, we can name them both.  
```

We could naiively emit BOTH in this case, which would cause a compile error in the int-footer, since we now have 2 explicit specializations by the same name. TBH, this would probably solve our access-specifier issue as well, but the error would come from the end of the host compiler.

Reply:
>So... how does this line work in situations where we have global-linkage but aren't allowed to spell the specialization constant's name? For example, if it is also a local variable inside a function?

This is already disallowed by the SYCL spec.  We say that `specialiation_id` variables must be declared at namespace scope (or class scope).

> OR, what if it is a 'private' static variable? We cannot spell that either.

This came up in the working group also.  We plan to say that a `specialization_id` variable declared in class scope must have public visibility.

> Another fun one! Anonymous namespaces!

This is an interesting case. Declaring a `specialization_id` variable in an anonymous namespace is OK so long as the name is unique.  The problem occurs only if the unqualified name is ambiguous.  I never thought about it before now, but I guess there is no way to fully qualify the name of a variable that is defined in an anonymous namespace?  If there is a way to do this, then that would solve the problem.

Reply:
Glad you guys have those 1st two solved.

>This is an interesting case. Declaring a specialization_id variable in an anonymous namespace is OK so long as the name is unique. The problem occurs only if the unqualified name is ambiguous. I never thought about it before now, but I guess there is no way to fully qualify the name of a variable that is defined in an anonymous namespace? If there is a way to do this, then that would solve the problem.

Yep, this is the interesting one.  You can name it unambiguously if you are IN the anonymous namespace, but by  this time we are already outside of it.  There isn't a way to spell it if there is a name collision.

Reply:
This still hasn't been closed on (anonymous namespaces being 'hidden').  At the moment, we're likely going to cause a host-error.

Reply:
I propose that we solve the anonymous namespace problem by adding this as a restriction in the SYCL spec.  I'm thinking of wording like this:

> If the `specialization_id` variable is declared in the anonymous namespace, its name must be unambiguous when referenced from the global namespace scope (i.e. it must not have the same name as some other variable declared in a named namespace).

We do not specifically say whether an implementation is required to diagnose any of the existing requirements for declaring `specialization_id` variables, so this would become another requirement that the implementation is not specifically required to check.  Of course, it would be better if our compiler did check for violations, but that is a quality of implementation thing.

Reply:
short-term we won't bother with the diagnostics, but at least any violations above are going to end up getting caught because the generated int-footer will be invalid.

Reply:
I will mark this as resolved since we have KhronosGroup/SYCL-Docs#133

Reply:
Actually, it may not be resolved.  There was some push back in the committee about disallowing applications where a `specialization_id` variable is declared in the unnamed (anonymous) namespace and also declared in the global namespace with the same name.  I thought about this some more, and there may be a solution.  I'd like to get your opinion.

The problematic case is something like this:

```
#include <sycl.hpp>

constexpr sycl::specialization_id<int> same_name{1};
namespace {
  constexpr sycl::specialization_id<int> same_name{2};
}

// ... More application code ...

// Footer with current design
namespace sycl::detail {

template<>
inline const char *get_spec_constant_symbolic_ID<::same_name>() {
  return "name for 'same_name' in global namespace";
}
template<>
inline const char *get_spec_constant_symbolic_ID<same_name>() {   // Compilation error here
  return "unique name for 'same_name' in unnamed namespace";
}

} // namespace sycl::detail

#include <CL/sycl/detail/spec_const_integration.hpp>
```

The code above would get a compilation error because the reference to `same_name` is ambiguous, and there is no way to remove the ambiguity by qualifying the name because there no fully qualified name for a variable in the unnamed namespace.

However, we could solve the problem like this:

```
#include <sycl.hpp>

constexpr sycl::specialization_id<int> same_name{1};
namespace {
  constexpr sycl::specialization_id<int> same_name{2};
}

// ... More application code ...

// Proposed new footer
namespace {
namespace sycl_detail {

template<auto &SpecName>
const char *get_spec_constant_symbolic_ID();

template<>
const char *get_spec_constant_symbolic_ID<::same_name>() {
  return "name for 'same_name' in global namespace";
}
template<>
const char *get_spec_constant_symbolic_ID<same_name>() {   // This is now unambiguous
  return "unique name for 'same_name' in unnamed namespace";
}

} // namespace sycl_detail
} // namespace (unnamed)

#include <CL/sycl/detail/spec_const_integration.hpp>
```

Note that the definition of `get_spec_constant_symbolic_ID()` is now inside the unnamed namespace.  As a result, the reference to unqualified `same_name` unambiguously resolves to the definition that is in the unnamed namespace.

The definition of `<CL/sycl/detail/spec_const_integration.hpp>` would look something like:

```
namespace sycl {

template<auto& SpecName>
void handler::set_specialization_constant(int value) {
  const char *name = sycl_detail::get_specialization_name<SpecName>();
  /* ... */
}

template<auto& SpecName>
int handler::get_specialization_constant() {
  const char *name = sycl_detail::get_specialization_name<SpecName>();
  /* ... */
}

} // namespace sycl
```

Notice how the definitions of `{set,get}_specialization_constant()` reference `sycl_detail::get_specialization_name()` from the unnamed namespace.

Disadvantages / ramifications of this design that I see are:

* Each translation unit will end up getting its own specialization of `{set,get}_specialization_constant()` for any particular `specialization_id` variable.  Even if two translation units reference the same variable, each TU will get its own definition because they reference `sycl_detail::get_specialization_name()`, which is defined with internal linkage in each TU.  I think this is not necessarily a problem, but it's something to keep in mind.

* We need to reserve a new namespace `sycl_detail`, which the implementation can use to avoid polluting the application's unnamed namespace.

Opinions?

If we don't see any showstopper here, I'd like to remove the wording in the SYCL 2020 spec that disallows a `specialization_id` variable from having the same name in both the unnamed namespace and also in the global namespace.  I think that makes the spec cleaner by removing a restriction that may seem somewhat arbitrary to users.

Reply:
I don't see the 'fix' you've done in the 2nd version, those both look identical.

Reply:
The difference is the namespace in which the specializations of `get_spec_constant_symbolic_ID()` are defined.  In the first example, they are defined in `::sycl::detail`.  In the second, they are defined in `(unnamed)::sycl_detail`.

Reply:
That only works for the 'global' anonymous namespace, doesn't it?  The issue ALSO appears for:

```
namespace MyNS {
constexpr sycl::specialization_id<int> same_name{1};
namespace {
  constexpr sycl::specialization_id<int> same_name{2};
}
}
```

Unless your proposal is to generate a specialization_id version for EVERY namespace in which we have something?  How would we end up finding it later?

Reply:
Yes, I did not consider the case of anonymous namespaces nested within other namespaces.  I think this can be solved, but it requires a runtime lookup.

```
#include <sycl.hpp>

constexpr sycl::specialization_id<int> same_name{1};
namespace {
  constexpr sycl::specialization_id<int> same_name{2};
}
namespace ns {
  constexpr sycl::specialization_id<int> same_name{3};
  namespace {
    constexpr sycl::specialization_id<int> same_name{4};
  }
}

// ... More application code ...

// Proposed new footer
namespace sycl::detail {

static specialization_id_map map[] = {
  {&::same_name, "name for 'same_name' in global namespace"},
  {&::ns::same_name, "name for 'same_name' in 'inner' namespace"}
};

} // namespace sycl::detail

namespace {
namespace sycl_detail {

static ::sycl::detail::specialization_id_map map[] = {
  {&same_name, "unique name for 'same_name' in global unnamed namespace"}
};

} // namespace sycl_detail
} // namespace (unnamed)

namespace ns {
namespace {
namespace sycl_detail {

static ::sycl::detail::specialization_id_map map[] = {
  {&same_name, "unique name for 'same_name' in 'ns' unnamed namespace"}
};

} // namespace sycl_detail
} // namespace (unnamed)
} // namespace ns

namespace sycl::detail {

static const char *get_specialization_name(const void *addr) {
  for (auto &e: ::sycl::detail::map) {
    if (e.addr == addr)
      return e.name;
  }
  for (auto &e: ::sycl_detail::map) {
    if (e.addr == addr)
      return e.name;
  }
  for (auto &e: ::ns::sycl_detail::map) {
    if (e.addr == addr)
      return e.name;
  }
  return nullptr;
}

} // namespace sycl::detail
```

The idea is to generate a series of arrays that map the address of each `specialization_id` variable to its string.  There is one array that maps all of the variables whose names can be fully qualified (i.e. not defined in an unnamed namespace).  We also generate one additional array for each unnamed namespace that contains a `specialzation_id` variable.  Finally, we generate a function with internal linkage that searches all of the arrays.

I think the definitions of `handler::{set,get}_specialization_constant()` can now be entirely in the `sycl.hpp` header (i.e. not included from the footer).  An example definition might look like:

```
// Included from <sycl.hpp> at the top of the translation unit

namespace sycl {
namespace detail {

struct specialization_id_map {
  const void *addr;
  const char *name;
};

static const char *get_specialization_name(const void *);

} // namespace detail


class specialization_id { /* ... */};

class handler {
 public:
  template<auto& SpecName>
  void set_specialization_constant(int value) {
    const char *name = detail::get_specialization_name(&SpecName);
    /* ... */
  }
};

} // namespace sycl
```

Reply:
That just moves the problem of the lookup to attempting to look up them map.  The function get_specialization_name now needs to be able to find all the 'maps', but it cannot do so due to anonymous namespaces.  Presumably we could  name that something other than map (some unique name?) and fully qualify that?  

This footer is getting... complicated to say the least.  @romanovvlad , what do you think about this?

Reply:
This seems to be getting uglier and VERY complicated.  Are we sure that supporting anonymous namespaces is important, and that it was intended by the SYCL 2020 standard?  Other than DPC++ how are other implementors of SYCL 2020 going to do this?

Reply:
> The function get_specialization_name now needs to be able to find all the 'maps', but it cannot do so due to anonymous namespaces. Presumably we could name that something other than map (some unique name?)

Of course, we could name all the `map` variables something unique, but I don't think it's necessary because we can disambiguate them via their namespace.  (Although there are unnamed namespaces here, each `map` is enclosed by some named namespace that makes the name unambiguous.)  I do have a prototype of this which works with both clang and gcc, so I'm reasonably confident about this part.

> This footer is getting... complicated

I agree.  I wanted to purse this to see if it would be reasonable to lift the restriction in the spec about requiring `specialiation_id` variables in the unnamed namespace to have unambiguous names.  Although I agree that the footer is getting more complicated, my bigger concern is about the extra runtime overhead imposed by the search in `get_specialization_name()`.

Reply:
> 
> 
> > The function get_specialization_name now needs to be able to find all the 'maps', but it cannot do so due to anonymous namespaces. Presumably we could name that something other than map (some unique name?)
> 
> Of course, we could name all the `map` variables something unique, but I don't think it's necessary because we can disambiguate them via their namespace. (Although there are unnamed namespaces here, each `map` is enclosed by some named namespace that makes the name unambiguous.) I do have a prototype of this which works with both clang and gcc, so I'm reasonably confident about this part.

Ok, I think I can buy that... I think because we control the containing namespace, we should be alright.

> 
> > This footer is getting... complicated
> 
> I agree. I wanted to purse this to see if it would be reasonable to lift the restriction in the spec about requiring `specialiation_id` variables in the unnamed namespace to have unambiguous names. Although I agree that the footer is getting more complicated, my bigger concern is about the extra runtime overhead imposed by the search in `get_specialization_name()`.

That is of course going to be somewhat significant, but at least it is just a O(N) search.

Reply:
> Are we sure that supporting anonymous namespaces is important, and that it was intended by the SYCL 2020 standard?

The anonymous namespace is a very normal C++ way of creating a variable that is private to a file.  I don't think we can exclude support of this entirely.  However, the problem is much more narrow.  The real problem occurs when the application uses the same variable name in the anonymous namespace and also in a parent namespace.  This is much less common.  I already proposed that we change the specification to disallow cases like this, but I got some pushback yesterday.  I think we need to do some due diligence to see if it is reasonable to support this case.  If we end up deciding it is unreasonable, we can go back to the committee and explain why we think the spec should disallow this case.

>  Other than DPC++ how are other implementors of SYCL 2020 going to do this?

I wanted to get feedback from our compiler team first on the proposal I floated above.  If we are open to this solution, I will touch base with Codeplay and see what they think.  I think their compiler will have a similar problem, and I'm not sure yet if they have a solution.

Reply:
Thanks Greg - Would you check with Codeplay concurrently?  If they are going to have a similar problem and don't yet have a solution, then I think it would be good to do see if we can get the SPEC changed to simplify both of the implementations.

Reply:
Yeah, please do that.  I know that @AlexeySachkov had wanted something out of the CFE to play with ASAP, and this is turning into a fairly significant effort.  Not undo-able obviously, but getting him something useful beyond what we have is going to be a lot of work.  I'd rather have CodePlay/the Spec locked down and non-negotiable before we start implementing that.

Reply:
> That is of course going to be somewhat significant, but at least it is just a O(N) search.

that can be even O(log2(N)), if all the maps are collected into std::map (backed by a balanced binary tree), which seems pretty OK to me. On the other hand, if we can't identify practical cases where ability to use spec const within unnamed space nested within a named namespace is really important, then why take even O(log2(N)) hit when we can avoid it entirely.

If such nesting is really important, we could specialize implementation so that it does runtime lookup only when FE finds such cases, otherwise use compile-time lookup.

Reply:
> On the other hand, if we can't identify practical cases where ability to use spec const within unnamed space nested within a named namespace is really important

I can imagine the following code, which uses this:
```
specializatoin_id<int> param;
namespace algorithm_1 {
  namespace {
    specialization_id<int> param;
  }
  int calculate(queue q, int p) {
    q.submit([&](handler &h) {
      h.set_specialization_constant<param>(p);
      h.parallel_for(/* ... */);
    }
  }
}
namespace algorithm_2 {
  namespace {
    specialization_id<int> param;
  }
  int calculate(queue q, int p) {
    q.submit([&](handler &h) {
      h.set_specialization_constant<param>(p);
      h.single_task(/* ... */);
    }
  }
}
```

The idea is that all kernels are written in the same style, i.e. they refer to the same `param` specialization constant. However, for each kernel it is a separate specialization constant, which might have different meaning/type/etc.

Not sure how useful this approach is, because personally, I would name those specialization constants differently if they really mean different things. Moreover, I still wasn't able to think of a situation where you also need some global specialization constant with the same name (I would add a prefix to it, clarifying that it was a global one instead of relying on `::` while accessing it - too easy to miss/forget).

However, I do agree that from the spec prospective the limitation we were trying to push seems weird:

> If the `specialization_id` variable is declared in the anonymous namespace, its name must be unambiguous when referenced from the global namespace scope (i.e. it must not have the same name as some other variable declared in a named namespace).

In all other places we say that specialization constant is identified by its *address* and not by its name/spelling.

Reply:
Note that the example you posted there @AlexeySachkov is actually fine, 'param' there are are all sufficiently uniquely named.

Reply:
> Note that the example you posted there @AlexeySachkov is actually fine, 'param' there are are all sufficiently uniquely named.

Oh, yes, got it. So, the proper example would be:
```c++
namespace algorithm {
  constexpr inline specializatoin_id<int> param; // user-visible
  namespace {
    constexpr specialization_id<int> param; // library internal
    int get_value(kernel_handler kh) {
      // we can't access internal param outside of anonymous namespace
      return kh.get_specialization_constant<param>() // internal
          + kh.get_specialization_constant<algorithm::param>(); // user-visible
    }
  }
  int calculate(queue q, int p) {
    q.submit([&](handler &h) {
      h.set_specialization_constant<param>(p); // set internal one 
      h.parallel_for([=](/* ... */, kernel_handler kh) {
        auto v = get_value(kh);
    }
  }
}
```

But I'm not sure about practical usefulness of that example: the only way to allow to set user-visible `param` outside of this code is to use `kernel_bundle`s, but then it won't be possible to set internal `param` as it is not accessible outside of this code (or it is better to say it shouldn't be if we consider this as a some real-life SYCL library). Unless, that kernel_bundle is passed to `calculate` and `build` is called within `calculate` to be able to set internal `param` as well.

Anyway, my point is:
- it is valid C++ and I don't see what would be a reasonable explanation why this doesn't work in SYCL for someone who is not familiar with implementation details
- however, this seems to be a some kind of corner-case and I can't think of a practical good looking example (not sure that shadowing variables with anonymous namespaces is a good practice)
- I need to take a look at the proposed solution one more time, but I hope that we would be able to detect in FE whether the reference to particular `specialization_id` is ambiguous or not (@erichkeane?) to decide whether or not we need to fallback to using runtime mapping
- also, the proposed mechanism doesn't change interface between components - it only complicates implementation of that interface for FE

If statements above make sense, then I suggest the following plan:
- I will incorporate this situation into design document
- We will implement it as lower priority unless proven that this is an important use-case

What do you think?

Reply:
I don't think we can avoid doing the 'map' lookup for everything, since the lookup is by address, so the CFE can't really tell what address is which.

And I think it does change the interface between components, because now the entry point is NOT a specialization, it is actually the 'lookup function' instead.

Reply:
> And I think it does change the interface between components, because now the entry point is NOT a specialization, it is actually the 'lookup function' instead.

But we can still pass the address to lookup through a template argument, right? I mean it is still possible to preserve the existing interface.

> I don't think we can avoid doing the 'map' lookup for everything, since the lookup is by address, so the CFE can't really tell what address is which.

Is it possible to collect all `specialization_id` declarations and group them by fully qualified names to see if some of them are the same? I imagine the following logic in FE:

```
1. collect all `specialization_id` declarations
2. group them by fully qualified name
3. for each `specialization_id`
  - if there the name is unique, generate a regular specialization within integration header
  - if the name is not unique, generate a specialization, which calls lookup function within it (or just put it implementation into  the specialization)
```

It is likely that I'm missing something, so, please correct me

Reply:
Looks like I was wrong about not being able to to avoid lookups for unambiguous ones.  Alexey pointed out in email that we can do:

```
template<auto &SpecName>
const char *get_spec_constant_symbolic_ID() {
  // here we generate map lookup for every ambiguous name
}

template<>
const char *get_spec_constant_symbolic_ID<nonAmbigiuosName>() {
  return “compile-time-known-id”;
} 
```

Reply:
Actually.... detecting ambiguity is going to be a lot of work in the CFE.  So we might need to do the 'map' look up for everything with an anonymous namespace in its decl context.